File: ExpenseTracker.java
import javax.swing.*;
import javax.swing.table.DefaultTableModel;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class ExpenseTracker extends JFrame {
    private DefaultTableModel tableModel;
    private JTable expenseTable;
    private JTextField expenseNameField;
    private JTextField amountField;
    private JTextField budgetField;
    private JLabel budgetLabel;
    private JLabel overspendLabel;
    private double budget = 0; // Default budget in Rs
    private double totalExpenses = 0.00;

    public ExpenseTracker() {
        setTitle("Expense Tracker");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(600, 400);
        setLocationRelativeTo(null);

        // Create the table model with Expense Name, Amount, and Delete columns
        tableModel = new DefaultTableModel(new Object[]{"Expense Name", "Amount (Rs)", "Delete"}, 0);
        expenseTable = new JTable(tableModel);

        // Add mouse listener to detect clicks on the Delete column
        expenseTable.addMouseListener(new MouseAdapter() {
            public void mouseClicked(MouseEvent e) {
                int column = expenseTable.getColumnModel().getColumnIndexAtX(e.getX());
                int row = e.getY() / expenseTable.getRowHeight();

                if (row < expenseTable.getRowCount() && row >= 0 && column == 2) {
                    removeExpense(row); // Remove the expense when "Delete" is clicked
                }
            }
        });

        // Input fields and add expense button
        JPanel inputPanel = new JPanel(new FlowLayout());
        expenseNameField = new JTextField(10);
        amountField = new JTextField(5);
        JButton addExpenseButton = new JButton("Add Expense");

        inputPanel.add(new JLabel("Expense Name:"));
        inputPanel.add(expenseNameField);
        inputPanel.add(new JLabel("Amount (Rs):"));
        inputPanel.add(amountField);
        inputPanel.add(addExpenseButton);

        // Budget setting panel
        JPanel budgetPanel = new JPanel(new FlowLayout());
        budgetField = new JTextField(5);
        JButton setBudgetButton = new JButton("Set Budget");
        budgetLabel = new JLabel("Total Expenses: Rs 0.00 / Budget: Rs " + budget);
        overspendLabel = new JLabel("");
        overspendLabel.setForeground(Color.RED);

        budgetPanel.add(new JLabel("Set Budget (Rs):"));
        budgetPanel.add(budgetField);
        budgetPanel.add(setBudgetButton);
        budgetPanel.add(budgetLabel);
        budgetPanel.add(overspendLabel);

        // Add action listener to "Add Expense" button
        addExpenseButton.addActionListener(e -> addExpense());

        // Add action listener to "Set Budget" button
        setBudgetButton.addActionListener(e -> setBudget());

        // Layout setup
        setLayout(new BorderLayout());
        add(inputPanel, BorderLayout.NORTH);
        add(new JScrollPane(expenseTable), BorderLayout.CENTER);
        add(budgetPanel, BorderLayout.SOUTH);
    }

    // Add a new expense to the table
    private void addExpense() {
        String expenseName = expenseNameField.getText();
        String amountText = amountField.getText();

        if (expenseName.isEmpty() || amountText.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Please enter both expense name and amount.", "Error", JOptionPane.ERROR_MESSAGE);
            return;
        }

        try {
            double amount = Double.parseDouble(amountText);
            totalExpenses += amount;

            // Add a new row with "Delete" action as "X"
            tableModel.addRow(new Object[]{expenseName, amount, "X"});
            updateBudgetLabel();
        } catch (NumberFormatException ex) {
            JOptionPane.showMessageDialog(this, "Please enter a valid amount.", "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    // Set a new budget based on user input
    private void setBudget() {
        String budgetText = budgetField.getText();

        if (budgetText.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Please enter a budget.", "Error", JOptionPane.ERROR_MESSAGE);
            return;
        }

        try {
            budget = Double.parseDouble(budgetText);
            updateBudgetLabel();
        } catch (NumberFormatException ex) {
            JOptionPane.showMessageDialog(this, "Please enter a valid budget amount.", "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    // Remove an expense from the table and update total expenses
    private void removeExpense(int row) {
        double amount = (double) tableModel.getValueAt(row, 1);
        totalExpenses -= amount;
        tableModel.removeRow(row); // Remove the row from the table model
        updateBudgetLabel();
    }

    // Update the label that shows the current budget and total expenses
    private void updateBudgetLabel() {
        budgetLabel.setText("Total Expenses: Rs " + totalExpenses + " / Budget: Rs " + budget);

        if (totalExpenses > budget) {
            double overspend = totalExpenses - budget;
            overspendLabel.setText("Overspent by Rs " + overspend);
            JOptionPane.showMessageDialog(this, "Warning: You have exceeded your budget by Rs " + overspend + "!", "Budget Warning", JOptionPane.WARNING_MESSAGE);
        } else {
            overspendLabel.setText(""); // Clear the overspend warning if under budget
        }
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new ExpenseTracker().setVisible(true));
    }
}

File: ExpenseTracker.py
import tkinter as tk
from tkinter import ttk, messagebox

class ExpenseTracker:
    def __init__(self, root):
        self.root = root
        self.root.title("Expense Tracker")
        self.root.geometry("600x400")

        self.budget = 0
        self.total_expenses = 0

        self.table_model = ttk.Treeview(self.root, columns=("Expense Name", "Amount (Rs)", "Delete"), show="headings")
        self.table_model.heading("Expense Name", text="Expense Name")
        self.table_model.heading("Amount (Rs)", text="Amount (Rs)")
        self.table_model.heading("Delete", text="Delete")
        self.table_model.pack(fill="both", expand=True)

        # Input fields and add expense button
        input_panel = tk.Frame(self.root)
        input_panel.pack(fill="x")
        tk.Label(input_panel, text="Expense Name:").pack(side="left")
        self.expense_name_field = tk.Entry(input_panel)
        self.expense_name_field.pack(side="left")
        tk.Label(input_panel, text="Amount (Rs):").pack(side="left")
        self.amount_field = tk.Entry(input_panel)
        self.amount_field.pack(side="left")
        add_expense_button = tk.Button(input_panel, text="Add Expense", command=self.add_expense)
        add_expense_button.pack(side="left")

        # Budget setting panel
        budget_panel = tk.Frame(self.root)
        budget_panel.pack(fill="x")
        tk.Label(budget_panel, text="Set Budget (Rs):").pack(side="left")
        self.budget_field = tk.Entry(budget_panel)
        self.budget_field.pack(side="left")
        set_budget_button = tk.Button(budget_panel, text="Set Budget", command=self.set_budget)
        set_budget_button.pack(side="left")
        self.budget_label = tk.Label(budget_panel, text="Total Expenses: Rs 0.00 / Budget: Rs 0")
        self.budget_label.pack(side="left")
        self.overspend_label = tk.Label(budget_panel, text="", fg="red")
        self.overspend_label.pack(side="left")

        delete_button = tk.Button(self.root, text="Delete Expense", command=self.delete_expense)
        delete_button.pack()

        self.root.mainloop()

    def add_expense(self):
        expense_name = self.expense_name_field.get()
        amount_text = self.amount_field.get()

        if not expense_name or not amount_text:
            messagebox.showerror("Error", "Please enter both expense name and amount.")
            return

        try:
            amount = float(amount_text)
            self.total_expenses += amount

            self.table_model.insert("", "end", values=(expense_name, amount, "Delete"))
            self.expense_name_field.delete(0, "end")
            self.amount_field.delete(0, "end")
            self.update_budget_label()
        except ValueError:
            messagebox.showerror("Error", "Please enter a valid amount.")

    def set_budget(self):
        budget_text = self.budget_field.get()

        if not budget_text:
            messagebox.showerror("Error", "Please enter a budget.")
            return

        try:
            self.budget = float(budget_text)
            self.update_budget_label()
        except ValueError:
            messagebox.showerror("Error", "Please enter a valid budget amount.")

    def update_budget_label(self):
        self.budget_label.config(text=f"Total Expenses: Rs {self.total_expenses:.2f} / Budget: Rs {self.budget:.2f}")

        if self.total_expenses > self.budget:
            overspend = self.total_expenses - self.budget
            self.overspend_label.config(text=f"Overspent by Rs {overspend:.2f}")
            messagebox.showwarning("Budget Warning", f"Warning: You have exceeded your budget by Rs {overspend:.2f}!")
        else:
            self.overspend_label.config(text="")

    def delete_expense(self):
        selected_item = self.table_model.focus()
        if selected_item:
            amount = self.table_model.item(selected_item, "values")[1]
            self.total_expenses -= float(amount)
            self.table_model.delete(selected_item)
            self.update_budget_label()
        else:
            messagebox.showerror("Error", "Please select an expense to delete.")

if __name__ == "__main__":
    root = tk.Tk()
    app = ExpenseTracker(root)
